// @serenity-js/core/lib/io/loader/ModuleLoader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModuleLoader = void 0;
// eslint-disable-next-line @typescript-eslint/no-var-requires
const Module = require('module'); // No type definitions available
const node_path_1 = __importDefault(require("node:path")); // eslint-disable-line unicorn/import-style
const Version_1 = require("../Version");
/**
 * Dynamically loads Node modules located relative to `cwd`.
 */
class ModuleLoader {
    cwd;
    useRequireCache;
    /**
     * @param {string} cwd
     *  Current working directory, relative to which Node modules should be resolved.
     * @param useRequireCache
     *  Whether to use Node's `require.cache`. Defaults to `true`.
     *  Set to `false` to force Node to reload required modules on every call.
     */
    constructor(cwd, useRequireCache = true) {
        this.cwd = cwd;
        this.useRequireCache = useRequireCache;
    }
    /**
     * Returns `true` if a given module is available to be required, false otherwise.
     *
     * @param moduleId
     *  NPM module id, for example 'cucumber' or '@serenity-js/core'
     */
    hasAvailable(moduleId) {
        try {
            return !!this.require(moduleId);
        }
        catch {
            return false;
        }
    }
    /**
     * Works like `require.resolve`, but relative to specified current working directory
     *
     * @param moduleId
     *  NPM module id, for example `cucumber` or `@serenity-js/core`
     *
     * @returns
     *  Path to a given Node.js module
     */
    resolve(moduleId) {
        const fromFile = node_path_1.default.join(this.cwd, 'noop.js');
        return Module._resolveFilename(moduleId, {
            id: fromFile,
            filename: fromFile,
            paths: Module._nodeModulePaths(this.cwd),
        });
    }
    /**
     * Works like `require`, but relative to specified current working directory
     *
     * @param moduleId
     */
    require(moduleId) {
        try {
            return require(this.cachedIfNeeded(this.resolve(moduleId)));
        }
        catch {
            return require(this.cachedIfNeeded(moduleId));
        }
    }
    cachedIfNeeded(moduleId) {
        if (!this.useRequireCache) {
            delete require.cache[moduleId];
        }
        return moduleId;
    }
    /**
     * Returns `Version` of module specified by `moduleId`, based on its `package.json`.
     *
     * @param moduleId
     */
    versionOf(moduleId) {
        return new Version_1.Version(this.require(`${moduleId}/package.json`).version);
    }
}
exports.ModuleLoader = ModuleLoader;
